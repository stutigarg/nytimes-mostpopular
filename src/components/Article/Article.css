.listView{
    padding-right:0;
    padding-left:0;
    height:100vh;
    overflow:auto
}
.detailedView{
    height:100vh;
}
@media (max-width: 767px) {
   .detailedView{
       height:50vh ;
       padding-right:0 ;
       padding-left:0
   }
 }
