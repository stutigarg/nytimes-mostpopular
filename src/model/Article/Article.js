class Article {

  /**
   * Constructs an own copy of article, with relatively few attributes, combined with custom attributes.
   *
   * @param article Article from NYTimes API
   */
  constructor(article) {
    this.id = article.id;
    this.url = article.url;
    this.title = article.title.replace('; ', '\n');
    this.summary = article.abstract.replace('; ', '\n');
    this.publishedOn = article.published_date;
    this.image = article.media[0]["media-metadata"][1].url;
    this.author = article.byline;

    this.selected = false;
  }

}

export default Article;
