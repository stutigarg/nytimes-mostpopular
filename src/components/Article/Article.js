import React, {Component} from 'react';
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import ArticleList from "./ArticleList";
import ArticleDetail from "./ArticleDetail";
import nytimes from "../../network";
import Loader from '../../util/Loader';
import {Config} from "../../.env";
import './Article.css';

class Article extends Component {

    state = {
        articles: [],
        selectedArticle: null,
        error: null
    }

    renderArticleDetails = (article) => {
        const currentArticle = {...this.state.selectedArticle};
        const articles = this.state.articles;
        articles.find(article => article.id === currentArticle.id).selected = false;
        article.selected = true;
        this.setState({articles: articles, selectedArticle: article})
    }

    componentDidMount() {
        this.getArticles(this.props.period);
    }

    getArticles = (period: Number) => {
        nytimes.articles(period)
            .then(result => {
                result[0].selected = true;
                this.setState({
                    articles: result,
                    selectedArticle: result[0]
                })
            })
            .catch(error => {
                this.setState({articles: [], error: error});
            });
    };

    render() {
        if (this.state.error) {
            throw this.state.error;
        }
        const loading = !this.state.error && this.state.articles.length === 0;
        const spinner = <Loader/>
        let detailedView = <ArticleDetail article={this.state.selectedArticle}/>
        let listView = loading ? spinner : <ArticleList articles={this.state.articles} onSelectHandler={this.renderArticleDetails}/>

        let data = 
            <div>
                <Container fluid={true}>
                    <Row>
                        <Col className="detailedView"sm={8}>{detailedView}</Col>
                        <Col className="listView" sm={4}>{listView}</Col>
                    </Row>
                </Container>
            </div>

        return data;
    }

}
Article.defaultProps = {
    period: Config.PERIOD
}

export default Article;
