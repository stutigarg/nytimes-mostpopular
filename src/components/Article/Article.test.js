import React from 'react';
import {configure, shallow} from 'enzyme';
import Article from './Article';
import Adapter from 'enzyme-adapter-react-16';
import renderer from "react-test-renderer";

configure({adapter: new Adapter()});

describe('<Article/>', () => {

  test('renders', async () => {
    const wrapper = await shallow(<Article />);
    expect(wrapper).toMatchSnapshot();
  });

});

describe('<Article/> with default valid period', () => {

  let articleComponent;

  beforeAll(async () => {
    articleComponent = await renderer.create(<Article/>).getInstance();
  });

  test('renders with articles', async () => {
    // Shouldn't be any error.
    expect(articleComponent.state.error).toBeNull();

    // Should render some articles
    expect(articleComponent.state.articles.length).not.toBe(0);

    // Should select 1st article by default
    expect(articleComponent.state.selectedArticle).toBe(articleComponent.state.articles[0]);
  });

  test('renders article details', async () => {
    // Should select 1st article by default
    const articles = articleComponent.state.articles;
    expect(articleComponent.state.selectedArticle).toBe(articles[0]);
    expect(articles.length).toBe(2);
    expect(articles[0].selected).toBe(true);
    expect(articles[1].selected).toBe(false);

    articleComponent.renderArticleDetails(articles[1]);
    expect(articles[0].selected).toBe(false);
    expect(articles[1].selected).toBe(true);
  });

});

describe('<Article/> with invalid period', () => {

  let articleComponent;

  beforeAll(async () => {
    articleComponent = await renderer.create(<Article period={0} />).getInstance();
  });

  test('renders with articles', async () => {
    // Should report an error.
    expect(articleComponent.state.error).not.toBeNull();

    // Shouldn't render any articles
    expect(articleComponent.state.articles.length).toBe(0);

    // And hence no selected article
    expect(articleComponent.state.selectedArticle).toBeNull();
  });

});