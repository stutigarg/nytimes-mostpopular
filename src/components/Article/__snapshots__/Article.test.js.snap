// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article/> renders 1`] = `ShallowWrapper {}`;
